Issue #397 - Drop servlet 3 and add support for Servlet 4 (Jakarta EE 8), Servlet 5 (Jakarta EE 9) and 6 (Jakarta EE 10) for v0.22

kevin-lee opened the pull request:
Issue #397 - Drop servlet 3 and add support for Servlet 4 (Jakarta EE 8), Servlet 5 (Jakarta EE 9) and 6 (Jakarta EE 10) for v0.22

- http4s-servlet no longer supports Servlet 3.
- Now, there are modules for Servlet 4, 5, and 6.
- The codebase was copied from https://github.com/http4s/http4s/tree/v0.22.15/servlet and then modified.

kevin-lee committed:
Issue #397 - Drop servlet 3 and add support for Servlet 4 (Jakarta EE 8), Servlet 5 (Jakarta EE 9) and 6 (Jakarta EE 10) for v0.22

- http4s-servlet no longer supports Servlet 3.
- Now, there are modules for Servlet 4, 5, and 6.
- The codebase was copied from https://github.com/http4s/http4s/tree/v0.22.15/servlet and then modified.

kevin-lee commented on build.sbt:
```diff
@@ -7,37 +7,147 @@ ThisBuild / developers := List(
+ThisBuild / crossScalaVersions := Seq("2.12.20", Scala213, "3.3.6")
 ThisBuild / scalaVersion := Scala213 // the default Scala
+
+ThisBuild / githubWorkflowJavaVersions --= List(JavaSpec.temurin("8"), JavaSpec.temurin("11"))
+ThisBuild / tlJdkRelease := Some(17)
```
We have a problem here.

Minimum Java versions:
* Servlet 4: Java 8
* Servlet 5: Java 8
* Servlet 6: Java 11

Source: https://en.wikipedia.org/wiki/Jakarta_EE

Jetty 12 requires Java 17.  
Although all `http4s-servlet` modules can be compiled with Java 11, testing needs Java 17 because Jetty 12 is used in the tests.

Since the `ci.yml` is generated by `githubWorkflowGenerate`, it's not quite possible to do fine-grained builds like testing with Java 17 and releasing with Java 8.

rossabaker replied:
`tlJdkRelease` sets which Java the emitted bytecode is compatible with, not the one used to build it.  So if the build runs on Java 17 and `tlJdkRelease` is `Some(8)`, it's probably fine.  Maybe it should be set in the project, not `ThisBuild`, for servlet 6.

kevin-lee replied:
OK, I'll give it a shot. I thought I had tried `tlJdkRelease is Some(8)`, but I could be wrong.

kevin-lee replied:
It doesn't seem to work. Whatever I put, I always get
```
      - name: Test
        run: sbt 'project ${{ matrix.project }}' '++ ${{ matrix.scala }}' test
```
So it doesn't check version for testing.

The following results in Java 8, 11 and 17 for 2.13 while Java 8 for 2.12 and 3 in build.
The problems are
* when testing, it doesn't check version so the test runs with 8 and 11.
* when publish it's Java 8 but it should be Java 11 for servlet 6.

You can have a look at [this branch](https://github.com/kevin-lee/http4s-servlet/commit/d08cae1bd5d3b09bba3fa0f6a97c68d649adb088).
```sbt
lazy val servlet5 = project
  .in(file("servlet5"))
  .settings(
    name := "http4s-servlet5",
    description := "Portable servlet implementation for http4s servers",
    githubWorkflowJavaVersions --= List(JavaSpec.temurin("8"), JavaSpec.temurin("11")),
    tlJdkRelease := Some(8),
    ...
)

lazy val servlet6 = project
  .in(file("servlet6"))
  .settings(
    name := "http4s-servlet6",
    description := "Portable servlet implementation for http4s servers",
    githubWorkflowJavaVersions --= List(JavaSpec.temurin("8"), JavaSpec.temurin("11")),
    tlJdkRelease := Some(11),
    ...
)
```
***
and the one below gives me only `17` for both build and publish.

You can check out [this branch](https://github.com/kevin-lee/http4s-servlet/commit/7784635823c09bfb997163f7155c9e24c56748a2).
```sbt
ThisBuild / githubWorkflowJavaVersions --= List(JavaSpec.temurin("8")

lazy val servlet5 = project
  .in(file("servlet5"))
  .settings(
    name := "http4s-servlet5",
    description := "Portable servlet implementation for http4s servers",
JavaSpec.temurin("11")),
    tlJdkRelease := Some(8),
    ...
  )

lazy val servlet6 = project
  .in(file("servlet6"))
  .settings(
    name := "http4s-servlet6",
    description := "Portable servlet implementation for http4s servers",
    tlJdkRelease := Some(11),
    ...
  )
```

kevin-lee replied:
Now, I'm replacing Jetty with Tomcat for testing, as Tomcat requires Java 8 for Servlets 4 and 5, and Java 11 for Servlet 6, which comes from the Servlet specs. So I think we can at least release it with Java 11. What do you think? Ideally, we could release each module with a different Java version, but I don’t know how to do that with the generated `ci.yaml`.

rossabaker replied:
I think there's a good chance you don't _need_ Servlet 6.  I think it's purely additive vs. Servlet 5.  The important thing is you have something that crosses the `javax.servlet` to `jakarta` chasm.

That said, it always feels good to compile and test against latest major versions, just to be sure.  Since these are new artifact IDs, I don't think there's any pressing need to support Java 8.  I'm fine with your strategy.

kevin-lee replied:
Got it, thanks! For me, Servlet 6 matters most since it's the only version where Jetty fixes "[java.lang.IllegalArgumentException: demand pending during load testing (#11694)](https://github.com/jetty/jetty.project/issues/11694)." 😅

I switched from Jetty to Tomcat for testing and had to make [this change](https://github.com/http4s/http4s-servlet/pull/399/files#r2179895452).

kevin-lee commented on build.sbt:
```diff
@@ -7,37 +7,147 @@ ThisBuild / developers := List(
+lazy val servlet4 = project
+  .in(file("servlet4"))
   .settings(
+    name := "http4s-servlet4",
```
I'm not sure if you like the name `http4s-servlet4`.
Would you prefer `-ee8` instead of the servlet version? Like `http4s-servlet-ee8` or `http4s-servlet-jee8`?

rossabaker replied:
I like servlet4, but it's not a strong preference.

kevin-lee commented on examples-servlet4/src/main/scala/com/example/Bootstrap.scala:
```diff
@@ -0,0 +1,54 @@
+import scala.concurrent.ExecutionContext
+
+@WebListener
+/** JettyPlugin from xsbt-web-plugin doesn't support Jetty 12.
+  * sbt-war, the successor of xsbt-web-plugin, doesn't support Jetty at all.
```
Given this, I'm not sure if the example module is even necessary.

rossabaker replied:
Yeah, this is for a module that glues together an old version of http4s with an old version of servlet.  This is not something we need to stretch ourselves to document examples of.

kevin-lee replied:
So, can I just remove all the example modules?

rossabaker replied:
Yes, I'd just remove them if they're causing any friction at all.

kevin-lee commented:
Once this PR is reviewed and merged, I'll update the other branches for http4s 0.23 with the same changes from this review.

kevin-lee commented on .github/workflows/ci.yml:
```diff
@@ -29,14 +29,9 @@ jobs:
         os: [ubuntu-22.04]
-        scala: [2.13, 3]
-        java: [temurin@8, temurin@11, temurin@17]
+        scala: [2.12, 2.13, 3]
```
Should we stop supporting version 2.12?

rossabaker replied:
http4s-core is only dropping support in the 1.x branch.  If it's causing pain, sure.  If it's not causing pain, I'd keep it.

kevin-lee commented on .github/workflows/ci.yml:
```diff
@@ -29,14 +29,9 @@ jobs:
-        scala: [2.13, 3]
-        java: [temurin@8, temurin@11, temurin@17]
+        scala: [2.12, 2.13, 3]
+        java: [temurin@17]
```
Please check out [this comment](https://github.com/http4s/http4s-servlet/pull/399/files#r2170779564).

rossabaker reviewed (commented):
Correct me if I'm wrong, but this is dropping support for Servlet 3, which was previously the _only_ supported version in the 0.22 line.  Since these are new artifact names, I guess that's fine.

kevin-lee commented:
> Correct me if I'm wrong, but this is dropping support for Servlet 3, which was previously the _only_ supported version in the 0.22 line. Since these are new artifact names, I guess that's fine.

@rossabaker Oh, my bad! Yeah, you're right. It's dropping Servlet 3 and adding 4, 5, and 6. I somehow thought it was already supporting 4. I'll fix the PR description and the commit message. Or do you want me to keep Servlet 3 as well?

kevin-lee committed:
Issue #397 - Replace Eclipse Jetty with Tomcat in servlet modules for servlet testing and update related tests

- Replace TestEclipseServer with TestTomcatServer in servlet modules
- Fix `NonBlockingServletIo` to read all available data by implementing recursive reading to consume all data in a single callback. This issue occurred with the Tomcat server, where data larger than the buffer size was not processed properly, leaving some data unconsumed.

rossabaker commented:
I don't think keeping Servlet 3 in this branch is strictly necessary.  0.22 is unsupported ... all this is mostly for you.  And if we decide to take pity on someone else who needs 0.22 and Servlet 3 later, we can revive off the branch I created before you started this.

kevin-lee commented on servlet4/src/main/scala/org/http4s/servlet/ServletIo.scala:
```diff
@@ -99,17 +99,66 @@ final case class NonBlockingServletIo[F[_]: Effect](chunkSize: Int) extends Serv
+                cb(rightSome(Chunk.empty))
+              } else {
+                finishReading(acc, totalSize)
+              }
+            } else {
+              /* Got some data */
+              if (len == chunkSize) {
+                val newAcc = acc :+ buf
+                val newTotalSize = totalSize + len
+                /* Got full chunk, continue reading as there might be more data is ready */
+                readRecursive(newAcc, newTotalSize)
+              } else {
+                val newAcc = acc :+ buf.take(len)
+                val newTotalSize = totalSize + len
+                /* If there are less data than chunkSize, it means available data are exhausted so finish reading. */
+                finishReading(newAcc, newTotalSize)
+              }
+            }
+          }
+
+        def finishReading(acc: Vector[Array[Byte]], totalSize: Int): Unit =
+          if (acc.isEmpty) {
+            /* No data was read. This may not happen in normal cases. */
+            cb(rightSome(Chunk.empty))
+          } else if (acc.size == 1) {
+            /* Single chunk, no need to combine */
+            cb(rightSome(Chunk.bytes(acc.head)))
+          } else {
+            /* Multiple chunks, combine them (already in correct order) */
+            val combined = new Array[Byte](totalSize)
+            var offset = 0
+            acc.foreach { chunk =>
+              val chunkLength = chunk.length
+              Array.copy(chunk, 0, combined, offset, chunkLength)
+              offset += chunkLength
+            }
+            cb(rightSome(Chunk.bytes(combined)))
+          }
+
+        readRecursive(Vector.empty, 0)
```
I had to make these changes because testing with Tomcat fails with the original code.

You can see the failed test here.
https://github.com/kevin-lee/http4s-servlet/actions/runs/16024200491/job/45208042585#step:10:128

<details>
<summary>Click to see the failed test</summary>

```
==> X org.http4s.servlet.AsyncHttp4sServletSuite.AsyncHttp4sServlet handle POST requests  0.578s munit.ComparisonFailException: /home/runner/work/http4s-servlet/http4s-servlet/servlet4/src/test/scala/org/http4s/servlet/AsyncHttp4sServletSuite.scala:91
90:
91:    post(server, "echo", contents).assertEquals(contents)
92:  }
values are not the same
=> Diff (- obtained, + expected)
 """,
-  "12 4095 SerXLyCygZTfTZTIJcGxiwRFvReUcoDroXGodQflTYskUBJNJvDsEgUldsCmziVEftfOFJlgcYEQtZvXMdZSxHnUGHzYOyksDIKelGKXWibqHyAeBjysqzPQSiJTJQVHDwvDfeuxXzZDYUJuZfMuatEaflYXEiHtMpxeCGAUlumCfyOTtShClJpCXAUiQqXqGgCuGWlaYLqLgJOAmZeizuqkYIhGsJslhklvgOIRaONvEwRffXwGRpgQkMvzIWXgRVyODNCDlrnJetsqSglFuVLivIqENyjBhXiqfEuBCfBOGBTOhYnjSredrgvddFJVmSbQpyZINcxlAvdcCvVtmfFnyUMTaiMYIGiabhLuqksulxvGhmuCZpOCFYuxgqVjfAvLXGHQQGSmIkYhYzcfvUqMExlYpRYUaaeBxQphRmIYRzxtxECLVZnzLJyPvuoLSirkfdOvzwMtFsVOBgbmNzxcJmXWRiVmDnOArKmVbNLoiUxgYLJhpvIDjXdiMBooptoKcWUecvmpeSsmiZzPBlyYFIBCvSdBYhzKLyQccEfxZkjuumhJXMyHDShieaDQLxoDWiDVFrUkmDheirXiRJGEVDbDNVjfznERHBJXsQzEUizQAnAKiweDyBVZxSUIEZxAWcZYwrmXCORusGuWntvYvyNjwuSIaJtFFnshbwbBBvqVUzwViXoOcIpJJQrLsSeACKJOfQUxjxOxuRCcBWSPNyXfRWURfVaoQQBrtePfPjKaLtsGYQNNxAKuCSECTsThkpXMBxaYZpOYTsOXFLwdtXZqnsZiQFzoobEsbjcjyFlnevEISnQPVVKiMlftqyogGkLhGjcIBVXXYlGMquMwlEMiSpDfcxYdCkZurOQMipCLYOtMxqMWpHHHLUYPzRFwmhiaqWbPeLbBnCrroPqlUOyctZHTPetIIIZaPttcLBlPFhVSmfzfnprlklIBgJrCvsWPyxWXDLtrNTnQzoqbmkZnnTmvFMXJzfuXeIKcWjyftVttcCoNRkcFOIhzgxkSRKtXYtsUujtUzSkikeMAjfUUpcSiDTgvQJuPEBeLxrJucyWWrIseWLFHcrxIrNqHMXlvGGLrehspcZQdxwohwLruocRVLGGPSxkPFVzXePeRyvzcPtZCerOawkvBBboqEjYSmjuMQxUWbwZqEMupTgyLdGCADdgtJMgaRzoosYpoDJkscBYWpgXTVHfJqpKCOthFyFSTisaGWsuRfnPRSVloGLVeGGKIUROVNpmPutXiMjlayzWyzlcWjppVWfwMIVIwhDBblafyUtvyYQlVHaQIVySMjSEJaFmfDgPEWMTYASaZVbWGEfhmfulJZXaYohqLnuMMaDCieuMAeIygsUkVtgJFMkaYgAISpDkjUKoZqOjaizfBoCbYLlDQweENpXllEeokDejpOEWYVjFnUjXAyhLeQKUIDpQXNbWDCOIewywxcVXkoRVAkTgXuHagbbmKQSzQqXWIJbnllMqmDrDsMIWrIhPiIVfVlAcNVJaHtLaRqlRuTBGZzciKGEFlquGnQHLryMnSgiVIzIBYeXnCvqobWoVhJwgScVeUSjHTFxLTGQykQloDLHbEGOAfiUEAyVHffieoaqSxNiqebrvFzHgLHIUCEhyHuDENrvFTtDpavaINZqmYYNaWJgGDrlPdADrZJIrqaXETtnSbzVTbheqbcFvtopLnMNafeOJyTCrflicHdQEiWOHrfLpoqsNhluHYZwjnGwGjJPDIfzHdHwyaQmqZtzqDwFQsCbQdYtZuVCDvkqThZlKAtXALlWuNFFIFVvssFWsoviXxmdHuroIhHQBzXYPuvIChfbEspZTqhrvIbpprNCzZBrnIOmMIktYEfAEJPcCbeflRLtUUYTwXlXMSWwfOuNdVxWnbtWtUuUWUMMnGzLqhpQzuMRQDWqAsZQKFtXUMOGYRYoIwSxqqmCjMeiowYGPikHEJkbcDJGjskFKiJWghvRiuyaDezLtulLWVRjbCfkqwwFxpdBHPqePyMPUDVfahMHfKeteZlAezghtQZETslfOIHzbwkNeKyEsYxDdOuJbIIuKwFEJJVCmzWuhdMiNxGSkMrMxDeoLeXTEcScNwWiNXcCqqXczzYUwPLswAxVVWWQLgaSqCIzKVychKgAgjpbcQjHDAYrnNNZjdnvpSWStIiHnXSXUyHjniVlnrKXtlbyEjtWpGJZGILxvgGMfvkTVRqnqiHwEvSvyEGoXZFyBmdPGwAvJugEzZOPWYboYXIhJiUZmGuzGyxxnVXUmMRyFNRfpibQsFavRNrUvPfmNCKSnllPiElUqumydreTgwUlisZGReeOeZErFtAxxnrrmJuIojQjxeRXozkLVhJLRUUfuJVgccYOaoFWAziBfGuxwXvKKSVqKVUzZBrqJTCiGgXWImroJzkjYPCMRmxFZzSvrREKtRPvtxYWzUbyUzQGVwaFeZnuybqPQZWQtBntgjPIPqrgjkpkcRmMIHdMOcuCflmUvIMLSaRWvIxhXXRTvmuObQsHTHQEYNsbMxciseKPvlLDjgymHcqbYfUXeTzzoknHJFathXGoGVJHknGphWfmDWEUjwsZTQRCoCmFyAATOcFUhUNYLMMUHLebLbdDwnILHHliaDBdCXtPjnJAHTtNtoQOwnksYKlCcRCDOyZHCWIYaYsEvjXYwzrvEtNBqagzJWBKGqcQuQTWVqjHhzpOhyKZpAWujtkOAKLoZTKYSdcCsXKbHXvDDnHAgUBuyIKInYqkjeqXedFAUddLvPROvbgwWKFKcbKwcEVAdNpZChuvtaXDIgwMEGKxpirOcWoJlzNNfhLbiVRkeIKkdivUnEhAnuBhngktFoMCoVYjuAYjbwsoAAyUDUVUHgBcUDdzPEIshXfTmWqzFWWNwieFxQVVkEHzUhlIxekjduaGSUjlgqbaIVqCobeLAZptakGiGKRqAdfsGxghLkAXWsOhlUMaCfdzbGkHaBhiEGKyaBMafsKIpryPoJPOVPmZnEhBtKjbOjVyntPpDarRJbTnokiAnwISxPjKOFrkpRaxPoSSElrjuXcnhSiaHWwntZrZfTKGHJPcPToIpjcwwIlctTEKdWxoFzpjjwLBEJyazhRILmEiTxfhNPZfIDDYiVBWqqCIZynDYnNOacIYXahwPakJKKPseOODggfcpMELkAZNFXEqIkPKHQXACSfnpMBntvjUhuLXVQwagwXICnFGgtcVcIGXqOidtFZWxyJBJEuwwvLbhdgZfdNqEfBJuzhxgiBEgfqriPfEMoiJhsrbjHypOPxUQxePngMWGcaoGFICWLJguoiTVxldGpvKVWsfYWhNYvEDTPMObYkPwUFqAlwScMjkcTkiApgZCSFDDFTLxXuurMBQxIkAQUbJGEWGzntKzWxlcpmmYXdtsNlIICVbKKBrbBqRFkZbUjYSYaESLWztlylwTEPUemsHmSXSnrvtcVujRtJUzedwvgCdZEJOxSwkTrzRrgPPPECvCceYXAqkKezyEOEYthGwPckGWMSKPUZeKrBeUxgwrgMNIAIYPctsidSudmnOGJzsTQAxxTpMsLKWClNWHNljHdnryyGrMrQlwSXirjqwrSDOCtDQOIyaOCdTTxvEONnVdiIMdpfAeGJFLLRnCLhtgOuBmyzebdVZCqVswsHITmsTehEYotxHxKYxAyhIERKzHOZKvKHFAOSredAyFjSRQbLfyAVrsmSNXkFNbGMeHDtaAAJNnnNLWfxzzBznsFNOwXTKMTnbWwEwDaCxAddRqDAyWvcbEUuapIVlBTjALusAUjUWOIWuFrFvRNXwjMGOmBrJGIXlAQLHGDYIPcgeOuURqgtGlJRQjeHLVEiujuzIaCPMpZInKKdSzfYxcXHKUGbPrlZgfNQVnIkAiZljsCvdHreIVpXWuZygKgcGIYEsudhyMFfixhTXOyyPqFzenfkbqRebBkhZvMPZZbbSldtOPDVUPpFxsyGEGklZwseRIPChsGjFDMVgEvipffVkZXbazNbYsEsAkIizcbxHnziDkJhmvRVFygfesbtaVhFWZBXRbFhZmridzaOtuNfCsUnIqc",
+  """12 4095 SerXLyCygZTfTZTIJcGxiwRFvReUcoDroXGodQflTYskUBJNJvDsEgUldsCmziVEftfOFJlgcYEQtZvXMdZSxHnUGHzYOyksDIKelGKXWibqHyAeBjysqzPQSiJTJQVHDwvDfeuxXzZDYUJuZfMuatEaflYXEiHtMpxeCGAUlumCfyOTtShClJpCXAUiQqXqGgCuGWlaYLqLgJOAmZeizuqkYIhGsJslhklvgOIRaONvEwRffXwGRpgQkMvzIWXgRVyODNCDlrnJetsqSglFuVLivIqENyjBhXiqfEuBCfBOGBTOhYnjSredrgvddFJVmSbQpyZINcxlAvdcCvVtmfFnyUMTaiMYIGiabhLuqksulxvGhmuCZpOCFYuxgqVjfAvLXGHQQGSmIkYhYzcfvUqMExlYpRYUaaeBxQphRmIYRzxtxECLVZnzLJyPvuoLSirkfdOvzwMtFsVOBgbmNzxcJmXWRiVmDnOArKmVbNLoiUxgYLJhpvIDjXdiMBooptoKcWUecvmpeSsmiZzPBlyYFIBCvSdBYhzKLyQccEfxZkjuumhJXMyHDShieaDQLxoDWiDVFrUkmDheirXiRJGEVDbDNVjfznERHBJXsQzEUizQAnAKiweDyBVZxSUIEZxAWcZYwrmXCORusGuWntvYvyNjwuSIaJtFFnshbwbBBvqVUzwViXoOcIpJJQrLsSeACKJOfQUxjxOxuRCcBWSPNyXfRWURfVaoQQBrtePfPjKaLtsGYQNNxAKuCSECTsThkpXMBxaYZpOYTsOXFLwdtXZqnsZiQFzoobEsbjcjyFlnevEISnQPVVKiMlftqyogGkLhGjcIBVXXYlGMquMwlEMiSpDfcxYdCkZurOQMipCLYOtMxqMWpHHHLUYPzRFwmhiaqWbPeLbBnCrroPqlUOyctZHTPetIIIZaPttcLBlPFhVSmfzfnprlklIBgJrCvsWPyxWXDLtrNTnQzoqbmkZnnTmvFMXJzfuXeIKcWjyftVttcCoNRkcFOIhzgxkSRKtXYtsUujtUzSkikeMAjfUUpcSiDTgvQJuPEBeLxrJucyWWrIseWLFHcrxIrNqHMXlvGGLrehspcZQdxwohwLruocRVLGGPSxkPFVzXePeRyvzcPtZCerOawkvBBboqEjYSmjuMQxUWbwZqEMupTgyLdGCADdgtJMgaRzoosYpoDJkscBYWpgXTVHfJqpKCOthFyFSTisaGWsuRfnPRSVloGLVeGGKIUROVNpmPutXiMjlayzWyzlcWjppVWfwMIVIwhDBblafyUtvyYQlVHaQIVySMjSEJaFmfDgPEWMTYASaZVbWGEfhmfulJZXaYohqLnuMMaDCieuMAeIygsUkVtgJFMkaYgAISpDkjUKoZqOjaizfBoCbYLlDQweENpXllEeokDejpOEWYVjFnUjXAyhLeQKUIDpQXNbWDCOIewywxcVXkoRVAkTgXuHagbbmKQSzQqXWIJbnllMqmDrDsMIWrIhPiIVfVlAcNVJaHtLaRqlRuTBGZzciKGEFlquGnQHLryMnSgiVIzIBYeXnCvqobWoVhJwgScVeUSjHTFxLTGQykQloDLHbEGOAfiUEAyVHffieoaqSxNiqebrvFzHgLHIUCEhyHuDENrvFTtDpavaINZqmYYNaWJgGDrlPdADrZJIrqaXETtnSbzVTbheqbcFvtopLnMNafeOJyTCrflicHdQEiWOHrfLpoqsNhluHYZwjnGwGjJPDIfzHdHwyaQmqZtzqDwFQsCbQdYtZuVCDvkqThZlKAtXALlWuNFFIFVvssFWsoviXxmdHuroIhHQBzXYPuvIChfbEspZTqhrvIbpprNCzZBrnIOmMIktYEfAEJPcCbeflRLtUUYTwXlXMSWwfOuNdVxWnbtWtUuUWUMMnGzLqhpQzuMRQDWqAsZQKFtXUMOGYRYoIwSxqqmCjMeiowYGPikHEJkbcDJGjskFKiJWghvRiuyaDezLtulLWVRjbCfkqwwFxpdBHPqePyMPUDVfahMHfKeteZlAezghtQZETslfOIHzbwkNeKyEsYxDdOuJbIIuKwFEJJVCmzWuhdMiNxGSkMrMxDeoLeXTEcScNwWiNXcCqqXczzYUwPLswAxVVWWQLgaSqCIzKVychKgAgjpbcQjHDAYrnNNZjdnvpSWStIiHnXSXUyHjniVlnrKXtlbyEjtWpGJZGILxvgGMfvkTVRqnqiHwEvSvyEGoXZFyBmdPGwAvJugEzZOPWYboYXIhJiUZmGuzGyxxnVXUmMRyFNRfpibQsFavRNrUvPfmNCKSnllPiElUqumydreTgwUlisZGReeOeZErFtAxxnrrmJuIojQjxeRXozkLVhJLRUUfuJVgccYOaoFWAziBfGuxwXvKKSVqKVUzZBrqJTCiGgXWImroJzkjYPCMRmxFZzSvrREKtRPvtxYWzUbyUzQGVwaFeZnuybqPQZWQtBntgjPIPqrgjkpkcRmMIHdMOcuCflmUvIMLSaRWvIxhXXRTvmuObQsHTHQEYNsbMxciseKPvlLDjgymHcqbYfUXeTzzoknHJFathXGoGVJHknGphWfmDWEUjwsZTQRCoCmFyAATOcFUhUNYLMMUHLebLbdDwnILHHliaDBdCXtPjnJAHTtNtoQOwnksYKlCcRCDOyZHCWIYaYsEvjXYwzrvEtNBqagzJWBKGqcQuQTWVqjHhzpOhyKZpAWujtkOAKLoZTKYSdcCsXKbHXvDDnHAgUBuyIKInYqkjeqXedFAUddLvPROvbgwWKFKcbKwcEVAdNpZChuvtaXDIgwMEGKxpirOcWoJlzNNfhLbiVRkeIKkdivUnEhAnuBhngktFoMCoVYjuAYjbwsoAAyUDUVUHgBcUDdzPEIshXfTmWqzFWWNwieFxQVVkEHzUhlIxekjduaGSUjlgqbaIVqCobeLAZptakGiGKRqAdfsGxghLkAXWsOhlUMaCfdzbGkHaBhiEGKyaBMafsKIpryPoJPOVPmZnEhBtKjbOjVyntPpDarRJbTnokiAnwISxPjKOFrkpRaxPoSSElrjuXcnhSiaHWwntZrZfTKGHJPcPToIpjcwwIlctTEKdWxoFzpjjwLBEJyazhRILmEiTxfhNPZfIDDYiVBWqqCIZynDYnNOacIYXahwPakJKKPseOODggfcpMELkAZNFXEqIkPKHQXACSfnpMBntvjUhuLXVQwagwXICnFGgtcVcIGXqOidtFZWxyJBJEuwwvLbhdgZfdNqEfBJuzhxgiBEgfqriPfEMoiJhsrbjHypOPxUQxePngMWGcaoGFICWLJguoiTVxldGpvKVWsfYWhNYvEDTPMObYkPwUFqAlwScMjkcTkiApgZCSFDDFTLxXuurMBQxIkAQUbJGEWGzntKzWxlcpmmYXdtsNlIICVbKKBrbBqRFkZbUjYSYaESLWztlylwTEPUemsHmSXSnrvtcVujRtJUzedwvgCdZEJOxSwkTrzRrgPPPECvCceYXAqkKezyEOEYthGwPckGWMSKPUZeKrBeUxgwrgMNIAIYPctsidSudmnOGJzsTQAxxTpMsLKWClNWHNljHdnryyGrMrQlwSXirjqwrSDOCtDQOIyaOCdTTxvEONnVdiIMdpfAeGJFLLRnCLhtgOuBmyzebdVZCqVswsHITmsTehEYotxHxKYxAyhIERKzHOZKvKHFAOSredAyFjSRQbLfyAVrsmSNXkFNbGMeHDtaAAJNnnNLWfxzzBznsFNOwXTKMTnbWwEwDaCxAddRqDAyWvcbEUuapIVlBTjALusAUjUWOIWuFrFvRNXwjMGOmBrJGIXlAQLHGDYIPcgeOuURqgtGlJRQjeHLVEiujuzIaCPMpZInKKdSzfYxcXHKUGbPrlZgfNQVnIkAiZljsCvdHreIVpXWuZygKgcGIYEsudhyMFfixhTXOyyPqFzenfkbqRebBkhZvMPZZbbSldtOPDVUPpFxsyGEGklZwseRIPChsGjFDMVgEvipffVkZXbazNbYsEsAkIizcbxHnziDkJhmvRVFygfesbtaVhFWZBXRbFhZmridzaOtuNfCsUnIqceALynOA
+""",
```
</details>

So if the request body exceeds the buffer size, only as many bytes as the buffer can hold are read.

e.g.) [this one](https://github.com/kevin-lee/http4s-servlet/blob/bbdb1728c8f2c880a64da2c92b822b7306947413/servlet4/src/test/scala/org/http4s/servlet/AsyncHttp4sServletSuite.scala#L100)
```scala
servletIo = NonBlockingServletIo[IO](4096)
```

This doesn't happen with Jetty, but it does with Tomcat.

So what I found is:  
- In `NonBlockingServletIo`, when `onDataAvailable()` ([#](https://github.com/kevin-lee/http4s-servlet/blob/94017bc3416a73af98939cff1e3bc2c2f2804fdb/servlet4/src/main/scala/org/http4s/servlet/ServletIo.scala#L127)) is called and `read()` ([#](https://github.com/kevin-lee/http4s-servlet/blob/94017bc3416a73af98939cff1e3bc2c2f2804fdb/servlet4/src/main/scala/org/http4s/servlet/ServletIo.scala#L101-L113)) runs,
- It reads bytes from the `InputStream` only up to the buffer size, then stops.
- The problem is, once this happens, `onAllDataRead()` ([#](https://github.com/kevin-lee/http4s-servlet/blob/94017bc3416a73af98939cff1e3bc2c2f2804fdb/servlet4/src/main/scala/org/http4s/servlet/ServletIo.scala#L139)) is called even if there are still bytes remaining in the `InputStream`.

rossabaker replied:
Yikes.  I guess it was good to test an alternate implementation!

rossabaker commented:
Since this is passing _and_ a huge PR, I'm going to merge and cut a milestone, and you can see if it works for you.  Once it does what you need, can cut a proper release.

rossabaker commented:
https://github.com/http4s/http4s-servlet/releases/tag/v0.22.0-M1

kevin-lee commented:
> Since this is passing _and_ a huge PR, I'm going to merge and cut a milestone, and you can see if it works for you. Once it does what you need, can cut a proper release.

@rossabaker Thank you so much for handling this so quickly. I’ll test and let you know the result soon.

kevin-lee commented:
@rossabaker I've opened a PR for `http4s-jetty` version `0.22`: https://github.com/http4s/http4s-jetty/pull/327
It would be good to have a milestone release of `http4s-jetty` so that I can test it with the service projects I have.
I'm also working on `http4s-servlet` for `http4s` version `0.23` now.

kevin-lee commented:
@rossabaker Testing is done, and I can confirm that it works. Could you please release it?

I'm still working on `0.24` and facing an issue with one test. I'll share the details later.

rossabaker commented:
Releasing as v0.22.0.

kevin-lee commented:
> Releasing as v0.22.0.

@rossabaker Thank you so much!
